[Team-04][Android][Linus_Jay][2주차 수요일 세번째 PR]

cherrylime69 opened the pull request:
## 🤷‍♂️ Description

<!-- 구현하고자 하는 기능에 대해 작성해 주세요. -->
- 필터 화면
  - 필터 조건 적용하기 구현
  - 조건이 적용된 이슈 목록 보여주기
  
- 이슈 목록 화면에서 클릭, 스와이프, 롱클릭 이벤트
  - swipe 를 통한 닫기
  - 숏클릭을 통한 상세화면 전환
  - 롱클릭을 통한 EditMode 전환
    - 이슈 선택 및 해제 (선택 개수 보여주기)
    - 선택 이슈들 일괄 닫기 및 삭제 구현
  - Label 보여주는 방식 수정
    - view 동적 생성 -> 아이템 마다  리사이클러뷰를 갖음

- 이슈 작성 화면
  - 커스텀 뷰를 활용하여 앱바 및 조건 적용바 구성
  - Markwon 라이브러리를 추가하여 미리 보기를 클릭하면 입력한 Issue body 가 마크다운 형식으로 표시되도록 구현
  - 임시적으로 FilterRepository 를 활용하여 스피너 값 불러오기 (나중에 따로 issueEditor 레포 생성 예정)
  - 제목을 작성해햐만 본문을 작성하도록 로직 작성. 제목 및 본문에 내용이 없으면 저장 아이템 활성화되지 않도록 구현
  - Markwon plugin 을 사용하여 이미지도 body에 넣으면 미리보기에서 보이도록 구현 완료

## 📷 Screenshots

<!--스크린샷으로 보여줄 수 있는 이미지가 있다면 첨부해주세요!-->
![Android Emulator - issu-tracker 2022-06-22 16-42-12](https://user-images.githubusercontent.com/95393311/174972935-65f06413-efe5-4698-937a-6d231eb69089.gif)


## 📝 기타 사항
- 추후 진행 상황
  -  이슈 작성 화면에서 popup menu 또는 window 표시하여 insertPhoto 명령 추가하기 
  - 이슈 목록 리스트에서 스와이프 및 롱클릭에서 버그가 조금 있어 수정 작업 진행 중

진행하면서 아직 질문 사항이 없어 추후에 생기면 바로 연락드리겠습니다! 날이 갑자기 더워졌는데 몸 조심하세욤 ㅎㅎ

<!--마지막으로 이슈 생성 시 우측의 옵션들을 체크했는지 확인해주세요!-->

cherrylime69 committed:
feat : 파이어스토어와 연동하여 filter 조건 데이터 동적 할당 완료

- 필터 화면으로 이동하면 파이어스토어에 저장된 데이터들을 할당하여 보여줌

cherrylime69 committed:
feat : 필터에 적용된 조건을 충족하는 이슈 목록을 보여주는 기능 구현

- 연산자 filter 와 any 를 활용하여 이전엔 로드한 이슈 목록 리스트의 필터하여 화면에 표시

shim506 committed:
Merge pull request #25 from shim506/feature/filter

firestore 와 연동하여 필터 조건 표시 및 필터링된 이슈 목록 보여주기

shim506 committed:
feat : 라벨 나타내는 방식 수성

동적 생성에서 -> 리사이클러뷰로

shim506 committed:
feat : Issue swipe 기능 구현

shim506 committed:
refactor : import 정리

shim506 committed:
feat : swipe 닫기 버튼을 통한 아이템 삭제 및 UI 갱신

shim506 committed:
Merge branch 'develop' into exp/touch_event

cherrylime69 committed:
Merge pull request #26 from shim506/exp/touch_event

swipe 를 통한 삭제 구현

shim506 committed:
fix : 클릭 버그 수정

-  스와이프 된 상태에서 닫기 눌렀을때로 수정
(스와이프 된 상태에서 다른 부분 클릭으로 삭제되는 버그 수정)

shim506 committed:
Merge branch 'exp/touch_event' of https://github.com/shim506/issue-tracker into exp/touch_event

cherrylime69 committed:
Merge pull request #27 from shim506/exp/touch_event

Swipe and Delete 버그 수정

shim506 committed:
fix : 빌드오류 수정

(과거 conflict 해결중 실수 )

shim506 committed:
fix : 스와이프 닫기 기능  수정 (삭제 -> 닫기(상태변화))

shim506 committed:
feat :  editMode 및 선택&선택 해제 구현

cherrylime69 committed:
feat : 이슈 작성 화면 구성 & 입력한 Text가 마크다운 형식으로 표시되도록 구현

- 커스텀 뷰를 활용하여 이슈 작성 화면 구성 완료
- Markwon 라이브러리를 추가하여 미리 보기를 클릭하면 입력한 Issue body 가 마크다운 형식으로 표시되도록 구현

shim506 committed:
feat : 선택 Issue들 지우기 및 닫기 구현

shim506 committed:
feat : editmode - Issue 선택 개수 반영

shim506 committed:
feat : Issue 숏클릭을 통산 상세페이지로의 젼환

상세페이지는 추후 진입점으로 쓰일 가능성이 있으므로 Activity 로 설정함

cherrylime69 committed:
Merge pull request #28 from shim506/feature/issue_edit_mode

Issue Event 마무리 구현 (+EditMode 포함)

cherrylime69 committed:
feat : 커스텀 조건바를 활용하여 특정 조건 적용하기, 저장 액션 아이템 활성화 분기 처리 및 이미지도 마크다운 형식으로 보여주기 구현

- 임시적으로 FilterRepository 를 활용하여 스피너 값 불러오기
- 제목을 작성해햐만 본문을 작성하도록 로직 작성. 제목 및 본문에 내용이 없으면 저장 아이템 활성화되지 않도록 구현
- Markwon plugin 을 사용하여 이미지 주소를 EditText 에 넣으면 미리보기에서 이미지가 보이도록 구현 완료

cherrylime69 committed:
Merge branch 'develop' into feature/issue-edit

cherrylime69 committed:
Merge pull request #30 from shim506/feature/issue-edit

이슈 작성 화면 및 일부 기능 구현 완료 (입력 가능, 마크다운 형식으로 표시 가능)

cherrylime69 committed:
fix : conflict 해결

cherrylime69 committed:
Merge pull request #31 from shim506/develop

[세번째 PR] 2주차 수요일 세번째 PR 요청

shim506 committed:
fix : 스와이프 상태 보존하지 못하는 버그 해결

- 다중 스와이프 후 닫기 할 경우 스와이프 상태 보존 못함
- 스와이프될 경우 스와이프 리스트에 추가하여 관리

shim506 committed:
fix : editmode 에서 스와이프가 가능한 버그 수정

renovatio0424 commented:
@cherrylime69 
걱정해주셔서 갬동이네요 ㅠㅠ 감사해요~ @cherrylime69 도 화이팅!

renovatio0424 commented on app/src/main/java/com/example/issu_tracker/data/Issue.kt:
```diff
@@ -1,12 +1,10 @@
-data class Label(val color: String = "", val content: String = "")
-
 data class IssueDto(
+    var id: String = "",
```
id 가 mutable 인게 잘 이해가 안되네요 ㅎㅎ;

renovatio0424 commented on app/src/main/java/com/example/issu_tracker/data/Label.kt:
```diff
@@ -0,0 +1,5 @@
+
+import java.io.Serializable
+
+data class Label(val color: String = "", val content: String = "") :Serializable
```
data model 에 color 가 있는건 좀 이상하네요 ㅎㅎ 
물론 필요한 경우도 있겠지만 label 의 타입에 따라 color 가 정해질거 같은데 ㅎㅎ

renovatio0424 commented on app/src/main/java/com/example/issu_tracker/data/repository/HomeRepositoryImpl.kt:
```diff
@@ -12,13 +17,47 @@ class HomeRepositoryImpl @Inject constructor(private val fireStore: FirebaseFire
+
+
+    override suspend fun deleteIssueList(list: List<Issue>) {
+        CoroutineScope(Dispatchers.IO).launch {
+            for (i in list) {
+                fireStore.collection(FIREBASE_COLLECTION_PATH).document(i.id).delete()
+            }
+        }.join()
```
`HomeRepositoryImpl` 안에서는 `Dispatchers.IO` 에서 작업이 되어야하는 상황이네요
이런 경우에는 `HomeRepositoryImpl` 에 `CoroutineScope` 인터페이스를 implement 해주면 좋은 상황이에요 ㅎㅎ 
한번 해보시길 ~

cherrylime69 replied:
![레노](https://user-images.githubusercontent.com/95393311/175191349-e5fb7c16-8039-4ee3-a9d6-0a4826328ab0.JPG)

말씀하신 피드백이 위의 이미지에서 보이는 코드처럼 구현해야 한다는 말씀이신가요?
만약 이렇게 구현하는 게 맞다면, 이와 같은 구현의 장점에 대해 말씀 또는 조언을 부탁드려도 될까요?

renovatio0424 commented on app/src/main/java/com/example/issu_tracker/ui/issue/IssueEditorFragment.kt:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    private fun addIssueBodyTextChangedListener() {
+        binding.etIssueEditorBody.addTextChangedListener(object : TextWatcher {
+            override fun beforeTextChanged(
+                inputText: CharSequence?,
+                start: Int,
+                count: Int,
+                after: Int
+            ) {
+            }
+
+            override fun onTextChanged(
+                inputText: CharSequence?,
+                start: Int,
+                before: Int,
+                count: Int
+            ) {
+            }
+
+            override fun afterTextChanged(inputText: Editable?) {
+                viewModel.inputIssueBodyText(inputText.toString())
+            }
+        })
+    }
```
참고로 `build.gradle` 파일에 이게 추가 되어야해요! 
`implementation 'androidx.core:core-ktx:1.0.0'`

cherrylime69 replied:
좋은 꿀팁 감사합니다 :D

renovatio0424 reviewed (approved):
수고하셨어요~ 
이해가 잘 안되는 코멘트들 있으면 언제든 연락주세요!

cherrylime69 committed:
feat : 이슈 작성 메뉴 기능 구현 완료

- ActionModeCallback 을 활용하여 롱클릭시 menu 창 활성화
(Default ActionMode 를 먼저 비활성화하고 롱클릭시 CustomActionMode 를 활성화하도록 구현)
- ClipBoard Manager 를 활용하여 Copy, Cut. Paste 기능 구현
- InsertPhoto 기능을 위해 Permission 요청과 권한이 있으면 외부 저장소에 접근하여 image 파일 path 불러오기 구현 완료

shim506 committed:
feat : 친구 목록 보여주기 기능 구현

cherrylime69 committed:
feat : 이슈 작성 메뉴 insertPhoto 기능 구현 완료

- 이미지 Uri 로부터 절대 경로 얻기
- 얻은 절대 경로를 firebase storage 에 업로드하기
- 업로드에 성공하면 업로드한 이미지 파일의 Uri 를 얻고 마크다운 형식으로 보여줄 수 있는 형식으로 변환하기
- 변환한 Uri 를 EditText 에 넣고 미리 보기를 클릭하면 마크다운 형식에 따라 이미지 보여주기

cherrylime69 committed:
feat : 이슈 작성 저장 누르면 새 이슈를 firebase 에 생성하기

- 이슈 제목, 본문(코멘트), 조건들의 값을 ViewModel에 전달
- 전달한 String 값과 일치하는 조건 데이터 클래스들을 저장
- 저장한 데이터 클래스, 이슈 제목 및 본문을 바탕으로 Issue 데이터 생성하여 Repository 에 전달
- Repo 에서 전달받은 값을 firebase 데이터 베이스에 저장

shim506 committed:
Merge pull request #33 from shim506/feature/issue-edit

이슈 작성 화면 및 기능 95% 구현 완료

shim506 committed:
feat : 로컬 레파지토리 구현 및 기존 레파지토리 아키텍처 수정

- Home, Filter 관련 레파지토리의 수정을 최소화 하고자 새로운 "Repository" 를 만들어 관리했고 내부적으로 remote,local 을 취급한다.

shim506 committed:
feat : 친구 목록 캐싱 후 가져오기 구현

shim506 committed:
feat : 친구 삭제 구현

shim506 committed:
Merge branch 'develop' into feature/room_account

cherrylime69 committed:
Merge pull request #34 from shim506/feature/room_account

내계정 친구 목록 부분 구현

cherrylime69 committed:
Merge pull request #35 from shim506/develop

[네번째 PR] 2주차 금요일 네번째 PR 요청
